From e233776ae60740ad7bbe01620eb3acc056001438 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?C=C3=A9dric=20VINCENT?= <cedric.vincent@st.com>
Date: Tue, 24 Mar 2020 17:18:21 +0100
Subject: [PATCH 1025/1695] scripts: render_font.py for Python 3.x
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

This reverts commit ef15dd07e960 ("TUI: remove unused
scripts/render_font.py") and upgrades render_font.py to python3.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 2f4e98f997496bef1149521319cb170397b44d57 scripts: render_font.py for Python 3.x]
Change-Id: I551cd8dc2a1a787a079fa3dc82fda7e0f1c56b1d
Signed-off-by: Lionel Debieve <lionel.debieve@foss.st.com>
Signed-off-by: Cédric VINCENT <cedric.vincent@st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/391777
Reviewed-by: Yannick FERTRE <yannick.fertre@foss.st.com>
Reviewed-by: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Domain-Review: Yannick FERTRE <yannick.fertre@foss.st.com>
Tested-by: Yannick FERTRE <yannick.fertre@foss.st.com>
---
 scripts/render_font.py | 141 +++++++++++++++++++++++++++++++++++++++++
 1 file changed, 141 insertions(+)
 create mode 100755 scripts/render_font.py

diff --git a/scripts/render_font.py b/scripts/render_font.py
new file mode 100755
index 000000000..7e1f07b4a
--- /dev/null
+++ b/scripts/render_font.py
@@ -0,0 +1,141 @@
+#!/usr/bin/env python3
+# SPDX-License-Identifier: BSD-2-Clause
+#
+# Copyright (c) 2016-2021, Linaro Limited
+#
+
+from wand.image import Image
+from wand.drawing import Drawing
+from wand.color import Color
+import sys
+
+
+def get_args():
+    from argparse import ArgumentParser
+
+    parser = ArgumentParser()
+    parser.add_argument('--font_file', required=True,
+                        help='Name of font file')
+    parser.add_argument('--font_size', type=int, default=40,
+                        help='Size of font')
+    parser.add_argument('--font_name', required=True,
+                        help='Name of font in program')
+    parser.add_argument('--out_dir', required=True,
+                        help='Out directory')
+    parser.add_argument('--verbose', default=False,
+                        help='Print informational messages')
+    return parser.parse_args()
+
+
+def c_hex_print(f, str):
+    n = 0
+    s = ""
+    for x in str:
+        if n % 8 == 0:
+            s += "\t"
+        else:
+            s += " "
+        # Hack to satisfy both Python 2.x and 3.x. In 2.x the variable x
+        # is string, while it in Python 3.x it's considered as a class
+        # int.
+        if sys.version_info > (3, 0):
+            s += "0x%02x," % x
+        else:
+            s += "0x%02x," % ord(x)
+        n = n + 1
+        if n % 8 == 0:
+            f.write(s + "\n")
+            s = ""
+    if n % 8 != 0:
+        f.write(s + "\n")
+        s = ""
+
+
+def write_comment(f):
+    f.write("/*\n * This file is auto generated with\n")
+    f.write(" *")
+    for x in sys.argv:
+        f.write(" " + x)
+    f.write("\n * do not edit.\n */\n")
+
+
+def main():
+    args = get_args()
+
+    draw = Drawing()
+    draw.font = args.font_file
+    draw.font_size = args.font_size
+
+    font_name = args.font_name
+    out_dir = args.out_dir
+
+    img_ref = Image(width=1000, height=1000)
+
+    if args.verbose:
+        print("Writing " + out_dir + "/" + font_name + ".c")
+    f = open(out_dir + "/" + font_name + ".c", 'w+')
+    write_comment(f)
+    f.write("#include \"font.h\"\n\n")
+
+    font_height = 0
+    range_first = 0x20
+    range_last = 0x7d
+    font_width = []
+    max_width = 0
+    for x in range(range_first, range_last + 1):
+        letter = chr(x)
+        metrics = draw.get_font_metrics(img_ref, letter)
+        text_height = int(round(metrics.text_height + 2))
+        if font_height == 0:
+            font_height = text_height
+        assert (font_height == text_height), "font height changed!"
+        if max_width == 0:
+            max_width = metrics.maximum_horizontal_advance + 2
+        assert (max_width == metrics.maximum_horizontal_advance + 2), \
+            "font advance width changed!"
+        text_width = int(round(metrics.text_width + 2))
+        font_width.append(text_width)
+        img = Image(width=text_width, height=text_height)
+        d = draw.clone()
+        d.text(0, int(metrics.ascender), letter)
+        d(img)
+
+        img.depth = 1
+
+        f.write("static const unsigned char ")
+        f.write("letter_" + str(hex(x)[2:]) + "[] = {\n")
+        c_hex_print(f, img.make_blob(format='A'))
+        f.write("};\n\n")
+        img.close()
+
+    f.write("static const struct font_letter letters[] = {\n")
+    for x in range(range_first, range_last + 1):
+        letter_var_name = "letter_" + str(hex(x)[2:])
+        f.write("\t{ " + letter_var_name + ", ")
+        f.write("sizeof(" + letter_var_name + "), ")
+        f.write(str(font_width[x - range_first]) + "},\n")
+    f.write("};\n\n")
+
+    f.write("const struct font font_" + font_name + " = {\n")
+    f.write("\t.first = " + str(hex(range_first)) + ",\n")
+    f.write("\t.last = " + str(hex(range_last)) + ",\n")
+    f.write("\t.letters = letters,\n")
+    f.write("\t.height = " + str(font_height) + ",\n")
+    f.write("\t.max_width = " + str(max_width) + ",\n")
+    f.write("};\n")
+    f.close()
+
+    if args.verbose:
+        print("Writing " + out_dir + "/" + font_name + ".h")
+    f = open(out_dir + "/" + font_name + ".h", 'w+')
+    write_comment(f)
+    f.write("#ifndef __" + font_name.upper() + "_H\n")
+    f.write("#define __" + font_name.upper() + "_H\n")
+    f.write("#include \"font.h\"\n")
+    f.write("extern const struct font font_" + font_name + ";\n")
+    f.write("#endif /*__" + font_name.upper() + "_H*/\n")
+    f.close()
+
+
+if __name__ == "__main__":
+    main()
-- 
2.47.2

